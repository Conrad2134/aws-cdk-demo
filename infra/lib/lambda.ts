import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as lambdaNode from '@aws-cdk/aws-lambda-nodejs';
import * as fs from 'fs';

export interface LambdaProps {
  entry: string;
  name: string;
}

export class Lambda extends cdk.Construct {
  public readonly handler: lambda.IFunction;

  constructor(scope: cdk.Construct, id: string, props: LambdaProps) {
    super(scope, id);

    // Make sure we can access the file, throw if we cannot.
    try {
      fs.accessSync(props.entry, fs.constants.F_OK);
    } catch {
      throw new Error(`Unable to access file at ${props.entry}`);
    }

    this.handler = new lambdaNode.NodejsFunction(this, `${id}StandardLambda`, {
      entry: props.entry,
      handler: 'handler',
      functionName: props.name,
      runtime: lambda.Runtime.NODEJS_14_X,
    });

    this.handler
      .metricErrors({
        period: cdk.Duration.minutes(1),
        statistic: cloudwatch.Statistic.SUM,
        unit: cloudwatch.Unit.COUNT,
      })
      .createAlarm(this, `${id}ErrorAlarm`, {
        alarmName: `${props.name}-error`,
        alarmDescription: `Alarm for errors in ${props.name}`,
        evaluationPeriods: 1,
        threshold: 1,
      });
  }
}
