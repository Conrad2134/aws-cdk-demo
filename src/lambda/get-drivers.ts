import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';

import drivers from '../data/drivers';

export async function handler(event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> {
  console.log('Getting drivers...');

  const { forceError } = event.queryStringParameters || {};

  if (forceError === 'true') {
    throw new Error('Error fetching drivers');
  }

  return {
    statusCode: 200,
    body: JSON.stringify({ count: drivers.length, data: drivers }),
    headers: {
      'access-control-allow-origin': '*',
      'access-control-allow-headers': 'Content-Type',
      'access-control-allow-methods': 'OPTIONS,GET',
    },
  };
}
