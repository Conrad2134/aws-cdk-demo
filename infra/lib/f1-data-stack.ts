import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as path from 'path';
import { Lambda } from './lambda';

interface F1ServiceStackProps extends cdk.StackProps {
  stage: 'dev' | 'prod';
}

export class F1DataStack extends cdk.Stack {
  public readonly url: cdk.CfnOutput;
  public readonly stage: cdk.CfnOutput;

  constructor(scope: cdk.Construct, id: string, props: F1ServiceStackProps) {
    super(scope, id, props);

    const getDrivers = new Lambda(this, `${props.stage}-GetDrivers`, {
      name: `${props.stage}-get-drivers`,
      entry: path.join(__dirname, '..', '..', 'src', 'lambda', 'get-drivers.ts'),
    });

    const api = new apigw.RestApi(this, `${props.stage}-Api`);
    api.root.addResource('drivers').addMethod('GET', new apigw.LambdaIntegration(getDrivers.handler));

    this.stage = new cdk.CfnOutput(this, 'Stage', { value: props.stage });
    this.url = new cdk.CfnOutput(this, 'Url', {
      value: api.url,
    });
  }
}
