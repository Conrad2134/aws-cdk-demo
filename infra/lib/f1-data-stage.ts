import { CfnOutput, Construct, Stage, StageProps } from '@aws-cdk/core';
import { F1DataStack } from './f1-data-stack';

interface F1DataStageProps extends StageProps {
  stage: 'dev' | 'prod';
}

/**
 * Deployable unit of web service app
 */
export class F1DataStage extends Stage {
  public readonly url: CfnOutput;

  constructor(scope: Construct, id: string, props: F1DataStageProps) {
    super(scope, id, props);

    const service = new F1DataStack(this, 'F1DataStack', { stage: props.stage });

    // Expose F1DataStack's output one level higher
    this.url = service.url;
  }
}
