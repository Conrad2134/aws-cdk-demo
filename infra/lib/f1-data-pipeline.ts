import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import { GitHubTrigger, ManualApprovalAction } from '@aws-cdk/aws-codepipeline-actions';
import { Construct, SecretValue, Stack, StackProps } from '@aws-cdk/core';
import { CdkPipeline, ShellScriptAction, SimpleSynthAction } from '@aws-cdk/pipelines';
import { F1DataStage } from './f1-data-stage';

/**
 * The stack that defines the application pipeline
 */
export class F1DataPipeline extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const sourceArtifact = new codepipeline.Artifact();
    const cloudAssemblyArtifact = new codepipeline.Artifact();

    const pipeline = new CdkPipeline(this, 'Pipeline', {
      // The pipeline name
      pipelineName: 'F1DataPipeline',
      cloudAssemblyArtifact,

      // Where the source can be found
      sourceAction: new codepipeline_actions.GitHubSourceAction({
        actionName: 'GitHub',
        output: sourceArtifact,
        oauthToken: SecretValue.secretsManager('github-token'),
        owner: 'Conrad2134',
        repo: 'aws-cdk-demo',
        branch: 'main',
        trigger: GitHubTrigger.POLL,
      }),

      // How it will be built and synthesized
      synthAction: SimpleSynthAction.standardNpmSynth({
        sourceArtifact,
        cloudAssemblyArtifact,
        installCommand: 'npm i -g npm@latest && npm ci',
        buildCommand: 'cd .. && npm ci',
        synthCommand: 'cd infra && npx cdk synth',
        subdirectory: 'infra',
      }),
    });

    // This is where we add the application stages
    const devStage = new F1DataStage(this, 'Dev', {
      env: { account: '194237782353', region: 'us-east-1' },
      stage: 'dev',
    });

    const devAppStage = pipeline.addApplicationStage(devStage);

    devAppStage.addActions(
      new ShellScriptAction({
        actionName: 'IntegrationTests',
        commands: ['npm i -g npm@latest', 'cd infra', 'npm ci', 'npm run test:integration'],
        additionalArtifacts: [sourceArtifact],
        runOrder: devAppStage.nextSequentialRunOrder(),
        useOutputs: {
          SERVICE_URL: pipeline.stackOutput(devStage.url),
        },
      }),
      new ManualApprovalAction({ actionName: 'Promote', runOrder: devAppStage.nextSequentialRunOrder() })
    );

    const prodStage = pipeline.addApplicationStage(
      new F1DataStage(this, 'Prod', {
        env: { account: '194237782353', region: 'us-east-1' },
        stage: 'prod',
      })
    );
  }
}
