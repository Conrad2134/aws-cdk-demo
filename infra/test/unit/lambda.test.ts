import { expect as expectCDK, haveResource, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Lambda } from '../../lib/lambda';
import * as path from 'path';

test('lambda is created with an alarm', () => {
  const stack = new cdk.Stack();

  new Lambda(stack, 'TestFunction', {
    name: 'test-function',
    entry: path.join(__dirname, 'utils', 'sample.ts'),
  });

  expectCDK(stack).to(haveResource('AWS::Lambda::Function'));
  expectCDK(stack).to(haveResource('AWS::CloudWatch::Alarm'));
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('error is thrown when an invalid entry path is supplied', () => {
  const stack = new cdk.Stack();

  expect(() => {
    new Lambda(stack, 'TestFunction', {
      name: 'test-function',
      entry: '../../hello.txt',
    });
  }).toThrowError('Unable to access file at ../../hello.txt');
});
