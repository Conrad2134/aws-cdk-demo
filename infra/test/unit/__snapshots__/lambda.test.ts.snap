// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambda is created with an alarm 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aArtifactHashEFAC242C": Object {
      "Description": "Artifact hash for asset \\"979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8a\\"",
      "Type": "String",
    },
    "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aS3BucketBE870156": Object {
      "Description": "S3 bucket for asset \\"979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8a\\"",
      "Type": "String",
    },
    "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aS3VersionKey6034C6D7": Object {
      "Description": "S3 key for asset version \\"979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8a\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestFunctionTestFunctionErrorAlarm5235A6CA": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm for errors in test-function",
        "AlarmName": "test-function-error",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "TestFunctionTestFunctionStandardLambda2CA2A9CB",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestFunctionTestFunctionStandardLambda2CA2A9CB": Object {
      "DependsOn": Array [
        "TestFunctionTestFunctionStandardLambdaServiceRoleA11353EE",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aS3BucketBE870156",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aS3VersionKey6034C6D7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters979f6fce7b86ca2b7d0d0bc2a62b49dc1eb28eb444e8781711946229cea76f8aS3VersionKey6034C6D7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "test-function",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestFunctionTestFunctionStandardLambdaServiceRoleA11353EE",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestFunctionTestFunctionStandardLambdaServiceRoleA11353EE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
